src/mscorlib/shared/System/Convert.cs

```diff
@@ -769,7 +771,9 @@ public static byte ToByte(object value, IFormatProvider provider)
 
         public static byte ToByte(bool value)
         {
-            return value ? (byte)Boolean.True : (byte)Boolean.False;
+            // Ideally we'd call BoolToByte(!!value) to normalize any true value to 1,
+            // but JIT optimizes !! away. The pattern below defeats this optimization.
```

Comment:
How can JIT optimize !! away? Is not it a bug?

Reply:
Hmm, haven't checked but both the JIT and the C# compiler tend to assume that `bool` is either 0 or 1. It's a bit of the bug since the IL spec doesn't actually guarantee that but then this behavior is old enough that it has become the norm.

Reply:
Checked, this is not JIT's doing. The C# compiler optimizes away `!!`.

Reply:
Yes, I was mistaken. The C# compiler optimizes away `!!`. I think this behavior is correct for the reason @mikedn calls out: the C# `!` operator is _not_ the same as the C `!` operator, and the user-observable side effect of double-notting a bool is to keep the original value. So this code isn't working around a bug as much as it is trying to make up for a missing language feature. I should reword the comment.

Additionally, as far as I can tell the JIT doesn't actually have a concept of a `!` operator; the `not` opcode is instead equivalent to one's complement (`~`). So putting two `not` opcodes into the stream doesn't have the desired effect either.

I can't think of a more efficient way of performing this normalization. One option would be to _not_ perform the normalization, but there's probably somebody _somewhere_ who's somehow passing a _true_ value other than 1 and relying on the existing logic.

Reply:
> Additionally, as far as I can tell the JIT doesn't actually have a concept of a ! operator; the not opcode is instead equivalent to one's complement (~). So putting two not opcodes into the stream doesn't have the desired effect either.

The usual way to negate a 0/1 bool is `xor 1`. Funnily enough, the C# compilers assumes 0/1 bools but fails to perform this particular optimization.

> One option would be to not perform the normalization, but there's probably somebody somewhere who's somehow passing a true value other than 1 and relying on the existing logic.

I doubt that not performing the normalization would have any meaningful impact considering that both the JIT (and AFAIR the JIT does some rather dubious things in this area) and the C# compilers perform various other optimizations assuming 0/1 bools. the ship has sailed a long time ago.

Reply:
> The usual way to negate a 0/1 bool is xor 1. Funnily enough, the C# compilers assumes 0/1 bools but fails to perform this particular optimization.

Didn't the compilers stop assuming that a few versions back because it was incorrect per ECMA? Currently if the C# compiler sees `!var` it will emit `ldarg [var]; ldc.i4.0; ceq;`, which JIT turns into `test ecx ecx; sete cl;`. This is correct for handling bools that aren't limited to 0 / 1.

Reply:
> This is correct for handling bools that aren't limited to 0 / 1.

Yes, and the same time the C# compile transforms `b1 && b2` into a bitwise `and`. That is only valid for 0/1 bools.

Reply:
> Yes, and the same time the C# compile transforms b1 && b2 into a bitwise and. That is only valid for 0/1 bools.

😒

OK, I'm convinced. I can remove the normalization logic and see what breaks.

Reply:
Maybe there should be a wider discussion involving C# people and JIT people. This is not how things should be done, with the IL spec saying one thing and different teams doing their own thing and doing it inconsistently.

There was an issue in the Roslyn repository about this that got closed with the team basically saying "we do what we want". I'm fine with that but if the JIT team and the FX teams also do what they want then the relevant part of the IL spec needs to be adjusted.

Reply:
@jaredpar who's the right set of folks to chat with?

Reply:
> Didn't the compilers stop assuming that a few versions back because it was incorrect per ECMA

No. The C# language only recognizes two boolean values: true and false. 

> This is not how things should be done, with the IL spec saying one thing and different teams doing their own thing and doing it inconsistently.

Don't think this is inconsistent. IL has a variety of features that C# does not support, booleans of values other than true or false are just one example. 

This has been true since C# 1.0.

Reply:
@jaredpar, I think the disconnect might be that both IL and C# say there are two boolean values: `true` and `false`. Additionally, both say that a `bool` is 1 byte.

However, the CLR explicitly defines it as `false is 0`; `true is not 0`. While C# does not clearly define the exact values that `true` and `false` are (just that there are only two values).

Since the C# spec doesn't explicitly call out what these values are, but the implementation generally handles it as `false is 0`; `true is not 0` but other times handles it as `false is 0`; `true is 1`, it leads to confusing behavior.

ECMA-335 definition:
> A CLI Boolean type occupies 1 byte in memory. A bit pattern of all zeroes denotes a value of
false. A bit pattern with any one or more bits set (analogous to a non-zero integer) denotes a
value of true. For the purpose of stack operations boolean values are treated as unsigned 1-byte
integers (§III.1.1.1).

C# Spec definition:
> The bool type represents boolean logical quantities. The possible values of type bool are true and false.

Reply:
> Don't think this is inconsistent. IL has a variety of features that C# does not support, booleans of values other than true or false are just one example.

I'm talking about the following inconsistency:
* `a && b` -> `a & b`
* `!!a` -> `a`
* `!a` -> `a == 0` (meh, `a xor 1` would generate a single instruction instead of 2 or 3)
* `a ^ true` -> `a == 0` (now that's just weird)

Granted, it would seem that using `a xor 1` doesn't work as well as expected due to the JIT having its own issues.

Reply:
> While C# does not clearly define the exact values that true and false are (just that there are only two values).

True. The C# spec does not explicitly define the value of `true` to any specific numerical value. However the C# compiler (as well as F#, VB, Cobol, etc ... every .NET language in existance) defines the literal `true` as the value `1` / `ldc.i4.1`. 

Scenarios where the literal `true` is imagined as any other numeric value are academic. 

> Since the C# spec doesn't explicitly call out what these values are, but the implementation generally handles it as false is 0; true is not 0

Incorrect. The compiler assumes booleans can only be 1 or 0 and emits it's logic based on that. Sure there are times where C#'s emitted code will happen to also work with a 2+ boolean value but that is just coincidence, not design. 

> The bool type represents boolean logical quantities. The possible values of type bool are true and false.

Note that the boolean type itself agrees with the C# compiler here on the numeric values of true and false. Both in the constants it defines internally and the values that are returned from TryParse.

Reply:
> Scenarios where the literal true is imagined as any other numeric value are academic.

Potentially academic, but still inline with the IL definition:
> A bit pattern with any one or more bits set (analogous to a non-zero integer) denotes a value of true.

I do think it would be desirable, and within the realm of reason, to have the C# spec explicitly list the values it recognizes as true/false; otherwise, two independent language implementations may do it differently (one may do 0/1, like Roslyn; and the other may do 0/not 0, like the IL spec).

Also, while I still think fixing the C# spec to explicitly have the same definition of `true` (that is, `true` is a non-zero bit pattern) would be nice. It could also be considered a breaking change in the scenarios that @mikedn listed above, since the code may now execute differently from before.

Reply:
The C# language specification is not specific to any particular runtime platform. The only observable consequence of placing something like that in the specification would be the behavior you get when you overlay a bool and a byte in unsafe code. Specifying that would be a peculiar departure from the normal situation in which the specification avoids telling you in detail the way unsafe code works.

Reply:
> Yes, and the same time the C# compile transforms b1 && b2 into a bitwise and. That is only valid for 0/1 bools.

@mikedn I was surprised to see that making this a bitwise operator is a Roslyn-specific behavior. Earlier versions of the C# compiler use short-circuiting logic. So this means technically there has already been a breaking change in the language, though the true number of people impacted is probably minimal. (Of course, there's still wonkiness with the other operators.)

Reply:
> I was surprised to see that making this a bitwise operator is a Roslyn-specific behavior

Note that the JIT does it too. But I'm not sure when it started doing it.